### PR TITLE
Master import excel files rba

### PR DESCRIPTION
**Pull request by rba-odoo**
Task: https://www.odoo.com/web#id=10792&view_type=form&model=project.task&menu_id=3942&action=327
pad: https://pad.odoo.com/p/r.U4xZYTHmXKS7NaBL

### CODE REVIEW

File: addons/base_import/static/src/js/import.js [hunk 1]
```diff
@@ -220,6 +220,11 @@ openerp.base_import = function (instance) {
             this.$('.oe_import_button, .oe_import_file_reload')
                     .prop('disabled', true);
             this.$el.addClass('oe_import_with_file');
+            if (!(this.$('input.oe_import_file').val().split('.').pop() == "csv")) {
+                this.$el.find('.oe_import_toggle').hide();
+            }else {
+                this.$el.find('.oe_import_toggle').show();
+            }
```
Comment (msh-odoo) [inline addons/base_import/static/src/js/import.js]: If file is kind of xls or xlsx or ods then there is no need of Separator and Quoting, but I think Encoding is still required

File: addons/base_import/models.py [hunk 2]
```diff
@@ -126,6 +133,33 @@ def get_fields(self, cr, uid, model, context=None,
...
+            rows = self._read_ods(record)
+        return rows
+
+    def _read_xls_xlsx(self, record):
+        book = xlrd.open_workbook(file_contents=record.file)
+        bk = book.sheet_by_index(0)
+        dict_list = []
+        for row_index in xrange(0, bk.nrows):
+            keys = [bk.cell(row_index, col_index).value for col_index in xrange(bk.ncols)]
+            dict_list.append(keys)
```
Comment (msh-odoo) [inline addons/base_import/models.py]: why not decoding it in utf8

File: addons/base_import/controllers.py [hunk 3]
```diff
@@ -7,12 +8,15 @@ class ImportController(Controller):
     @route('/base_import/set_file')
     def set_file(self, req, file, import_id, jsonp='callback'):
         import_id = int(import_id)
+        if file.filename[-3:] == "ods":
```
Comment (xmo-odoo) [inline addons/base_import/controllers.py]: Wouldn't it be a better idea to switch based on the file's content_type? That's what is being used elsewhere.

File: addons/base_import/models.py [hunk 4]
```diff
@@ -2,6 +2,13 @@
 import itertools
 import logging
 import operator
+from odsreader import *
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: No star-imports. Also incorrectly sorted  imports

File: addons/base_import/models.py [hunk 5]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+        if record.file_type == 'text/csv':
+            rows = self._read_csv(record, options)
+        elif record.file_type == 'application/vnd.ms-excel' or record.file_type == 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet':
+            if xlrd is None:
+                raise UserError(_(" XLRD library not found. XLS/XLSX file would not import."))
+            else:
+                rows = self._read_xls_xlsx(record, options)
+        elif record.file_type == 'application/vnd.oasis.opendocument.spreadsheet':
+            rows = self._read_ods(record, options)
+        return rows
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Why is there a UserError raised if trying to import an XLS/XLSX file but no information when trying to import an unkwown data type, just an empty list of rows?

File: addons/base_import/models.py [hunk 6]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
         # TODO: cache on model?
         return fields
 
+    def _read_import_file(self, record, options):
+        rows = []
+        if record.file_type == 'text/csv':
+            rows = self._read_csv(record, options)
+        elif record.file_type == 'application/vnd.ms-excel' or record.file_type == 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet':
+            if xlrd is None:
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Why not just `if not xlrd`?

File: addons/base_import/models.py [hunk 7]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
         # TODO: cache on model?
         return fields
 
+    def _read_import_file(self, record, options):
+        rows = []
+        if record.file_type == 'text/csv':
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: A dispatch dict (content_type -> handler function) or at least a dispatch method (takes content_type, returns handler method) would probably make sense, and make for easier overloading.

File: addons/base_import/models.py [hunk 8]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+        return rows
+
+    def _read_xls_xlsx(self, record, options):
+        book = xlrd.open_workbook(file_contents=record.file)
+        bk = book.sheet_by_index(0)
+        keys = []
+        for row_index in range(0, bk.nrows):
+            keys.append([])
+            for col_index in range(bk.ncols):
+                if bk.cell_type(row_index, col_index) == 3:
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Use the library's named constants, not magic numbers.

File: addons/base_import/models.py [hunk 9]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+                rows = self._read_xls_xlsx(record, options)
+        elif record.file_type == 'application/vnd.oasis.opendocument.spreadsheet':
+            rows = self._read_ods(record, options)
+        return rows
+
+    def _read_xls_xlsx(self, record, options):
+        book = xlrd.open_workbook(file_contents=record.file)
+        bk = book.sheet_by_index(0)
+        keys = []
+        for row_index in range(0, bk.nrows):
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: lower bound unnecessary.

File: addons/base_import/models.py [hunk 10]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+            keys.append([])
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Why not define a `row = []` and just add the row to whatever `keys` is at the end of the loop?

File: addons/base_import/models.py [hunk 11]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+                        keys[-1].append(datetime.datetime(* \
+                        (xlrd.xldate_as_tuple(bk.cell_value(row_index, col_index), \
+                            book.datemode))).strftime('%Y-%m-%d %H:%M:%S'))
+                elif bk.cell_type(row_index, col_index) == 2:
+                    keys[-1].append(str(bk.cell_value(row_index, col_index)))
+                elif bk.cell_type(row_index, col_index) == 4:
+                    if bk.cell_value(row_index, col_index) == 1:
+                        keys[-1].append('True')
+                    elif bk.cell_value(row_index, col_index) == 0:
+                        keys[-1].append('False')
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: `str(bool(bk.cell_value(row_index, col_index))`?

File: addons/base_import/models.py [hunk 12]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+
+    def _read_xls_xlsx(self, record, options):
+        book = xlrd.open_workbook(file_contents=record.file)
+        bk = book.sheet_by_index(0)
+        keys = []
+        for row_index in range(0, bk.nrows):
+            keys.append([])
+            for col_index in range(bk.ncols):
+                if bk.cell_type(row_index, col_index) == 3:
+                    date_decimal = str(bk.cell_value(row_index, col_index)-int(bk.cell_value(row_index, col_index)))[1:].split('.')[1]
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: If you want only the decimal part of the number, why not just use `% 1`? e.g.
```python
>>> 1.3 % 1
0.30000000000000004
>>> 39217.54 % 1
0.5400000000008731
```

File: addons/base_import/tests/test_cases.py [hunk 13]
```diff
@@ -253,6 +256,103 @@ def test_success(self):
         # Ensure we only have the response fields we expect
         self.assertItemsEqual(result.keys(), ['matches', 'headers', 'fields', 'preview'])
 
+    def test_xls_success(self):
+        try:
+            import xlrd as xlrd
+        except ImportError:
+            return True
```
Comment (xmo-odoo) [inline addons/base_import/tests/test_cases.py]: Use [`unittest.skipIf`](https://docs.python.org/2/library/unittest.html#unittest.skipIf)

File: addons/base_import/static/src/js/import.js [hunk 14]
```diff
@@ -206,10 +206,14 @@ openerp.base_import = function (instance) {
 
         //- File & settings change section
         onfile_loaded: function () {
+            if (!(this.$('input.oe_import_file').val().split('.').pop() == "csv")) {
+                this.$el.find('.oe_import_toggled').children().nextAll().hide()
```
Comment (xmo-odoo) [inline addons/base_import/static/src/js/import.js]: * Use `this.$()`
* `.children().slice(1)` is a much better and clearer way to select all-but-the-first children of the reference node

File: addons/base_import/static/src/js/import.js [hunk 15]
```diff
@@ -206,10 +206,14 @@ openerp.base_import = function (instance) {
...
+            }else {
```
Comment (xmo-odoo) [inline addons/base_import/static/src/js/import.js]: spacing

File: addons/base_import/models.py [hunk 16]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+                    if date_decimal == '0':
+                        keys[-1].append(datetime.datetime(* \
+                        (xlrd.xldate_as_tuple(bk.cell_value(row_index, col_index), \
+                            book.datemode))).strftime('%Y-%m-%d'))
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Might be simpler to just use time.strftime, as the second parameter can be a timetuple.

File: addons/base_import/models.py [hunk 17]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+                    keys[-1].append(str(bk.cell_value(row_index, col_index)))
+                elif bk.cell_type(row_index, col_index) == 4:
+                    if bk.cell_value(row_index, col_index) == 1:
+                        keys[-1].append('True')
+                    elif bk.cell_value(row_index, col_index) == 0:
+                        keys[-1].append('False')
+                else:
+                    keys[-1].append(bk.cell_value(row_index, col_index))
+        encoding = options.get('encoding', 'utf-8')
+        dict_list = itertools.imap(lambda row: [item.decode(encoding) for item in row], keys)
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: The data is already decoded, this step makes no sense: store unicode values in the first place (so don't reencode text cells badly)

File: addons/base_import/models.py [hunk 18]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+        return dict_list
+
+    def _read_ods(self, record, options):
+        rows = odsreader(record.file)
+        keys = [row for row in rows]
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: also known as `keys = rows`

File: addons/base_import/static/src/js/import.js (hunk 1 above)
Comment (xmo-odoo) [inline addons/base_import/static/src/js/import.js]: Nope, as far as I can tell xlrd always returns unicode, and the expat parser underlying the ods reader also always returns unicode data (unless reconfigured not to by setting `returns_unicode` to `False`)

File: addons/base_import/odsreader.py [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        if self.__iter_cnt == len(self.__rows):
+            raise StopIteration
+        else:
+            self.__iter_cnt = self.__iter_cnt + 1
+            return self.__rows[self.__iter_cnt - 1]
```
Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: reimplement `__iter__` with a generator, so readers can be reused

File: addons/base_import/models.py [hunk 20]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+        encoding = options.get('encoding', 'utf-8')
+        dict_list = itertools.imap(lambda row: [item.decode(encoding) for item in row], keys)
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: As far as I understand the rest of the code, ODS text is already decoded. So you can just return `odsreader(record.file)`

File: addons/base_import/models.py [hunk 21]
```diff
@@ -126,6 +134,56 @@ def get_fields(self, cr, uid, model, context=None,
...
+                            book.datemode))).strftime('%Y-%m-%d %H:%M:%S'))
+                elif bk.cell_type(row_index, col_index) == 2:
+                    keys[-1].append(str(bk.cell_value(row_index, col_index)))
+                elif bk.cell_type(row_index, col_index) == 4:
+                    if bk.cell_value(row_index, col_index) == 1:
+                        keys[-1].append('True')
+                    elif bk.cell_value(row_index, col_index) == 0:
+                        keys[-1].append('False')
+                else:
+                    keys[-1].append(bk.cell_value(row_index, col_index))
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: type 5 `XL_CELL_ERROR` probably shouldn't be handled that way.

File: addons/base_import/static/src/js/import.js [hunk 22]
```diff
@@ -206,10 +206,14 @@ openerp.base_import = function (instance) {
 
         //- File & settings change section
         onfile_loaded: function () {
+            if (!(this.$('input.oe_import_file').val().split('.').pop() == "csv")) {
```
Comment (xmo-odoo) [inline addons/base_import/static/src/js/import.js]: Should use the file's inferred or configured content type.

File: addons/base_import/models.py (hunk 16 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: actually xlrd.xldate_as_tuple will return tuple in 6 digits for example (2015, 3, 5, 8, 0, 0), while time.strftime requires tuple like this (2015, 3, 5, 16, 14, 17, 978221).

**Comment by rba-odoo** [comment]
@xmo-odoo resolved your comments, can you have a look at it.?

File: addons/base_import/models.py [hunk 23]
```diff
@@ -126,6 +134,66 @@ def get_fields(self, cr, uid, model, context=None,
...
+            raise UserError(_("Not a compatible file format. It should be XLS, XLSX or ODS file"))
+
+    def _read_xls(self, record, options):
+        cell_type_index = {'XL_CELL_EMPTY': 0,
+                            'XL_CELL_TEXT': 1,
+                            'XL_CELL_NUMBER': 2,
+                            'XL_CELL_DATE': 3,
+                            'XL_CELL_BOOLEAN': 4,
+                            'XL_CELL_ERROR': 5,
+                            'XL_CELL_BLANK': 6}
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: XLRD already provides all the relevant constants...

File: addons/base_import/models.py [hunk 24]
```diff
@@ -126,6 +134,66 @@ def get_fields(self, cr, uid, model, context=None,
         # TODO: cache on model?
         return fields
 
+    def _file_type_handler(self, file_type, record, options):
+        file_type_dict = {'text/csv': 'csv',
+                    'application/vnd.ms-excel': 'xls',
+                    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet':'xlsx',
+                    'application/vnd.oasis.opendocument.spreadsheet':'ods'
+                    }
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Should be a module-level constant

File: addons/base_import/models.py [hunk 25]
```diff
@@ -126,6 +134,66 @@ def get_fields(self, cr, uid, model, context=None,
...
+                    if bk.cell_value(row_index, col_index) in xlrd.error_text_from_code:
+                        return "Error: (%s)" % xlrd.error_text_from_code[bk.cell_value(row_index, col_index)]
+                    return "Unknown Erro: (%r)" % bk.cell_value(row_index, col_index)
+                else:
+                    keys.append(bk.cell_value(row_index, col_index))
+            rows.append(keys)
+        return iter(rows)
+
+    def _read_xlsx(self, record, options):
+        return self._read_xls(record, options)
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: `_read_xlsx = _read_xls` would work just as well in that case

File: addons/base_import/models.py [hunk 26]
```diff
@@ -126,6 +134,66 @@ def get_fields(self, cr, uid, model, context=None,
...
+
+    def _read_ods(self, record, options):
+        return iter(odsreader(record.file))
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Is there really a need for the `iter` call?

File: addons/base_import/models.py [hunk 27]
```diff
@@ -126,6 +134,66 @@ def get_fields(self, cr, uid, model, context=None,
...
+                    elif bk.cell_value(row_index, col_index) == 0:
+                        keys.append('False')
+                elif bk.cell_type(row_index, col_index) == cell_type_index['XL_CELL_ERROR']:
+                    if bk.cell_value(row_index, col_index) in xlrd.error_text_from_code:
+                        return "Error: (%s)" % xlrd.error_text_from_code[bk.cell_value(row_index, col_index)]
+                    return "Unknown Erro: (%r)" % bk.cell_value(row_index, col_index)
+                else:
+                    keys.append(bk.cell_value(row_index, col_index))
+            rows.append(keys)
+        return iter(rows)
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: Is there really a need for the `iter` call? In fact I think you could get rid of the `rows` list and just `yield keys` inside the loop

File: addons/base_import/models.py [hunk 28]
```diff
@@ -126,6 +134,66 @@ def get_fields(self, cr, uid, model, context=None,
...
+                            book.datemode))).strftime('%Y-%m-%d %H:%M:%S'))
+                elif bk.cell_type(row_index, col_index) == cell_type_index['XL_CELL_BOOLEAN']:
+                    if bk.cell_value(row_index, col_index) == 1:
+                        keys.append('True')
+                    elif bk.cell_value(row_index, col_index) == 0:
+                        keys.append('False')
+                elif bk.cell_type(row_index, col_index) == cell_type_index['XL_CELL_ERROR']:
+                    if bk.cell_value(row_index, col_index) in xlrd.error_text_from_code:
+                        return "Error: (%s)" % xlrd.error_text_from_code[bk.cell_value(row_index, col_index)]
+                    return "Unknown Erro: (%r)" % bk.cell_value(row_index, col_index)
```
Comment (xmo-odoo) [inline addons/base_import/models.py]: I don't think the caller expects an error to be returned, so this should probably be a raised exception.

File: addons/base_import/odsreader.py [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+
+        p.StartElementHandler = self.__start_element__
+        p.EndElementHandler = self.__end_element__
+        p.CharacterDataHandler = self.__char_data__
+        # parse the data
+        p.Parse(xmldata, True)
+
+    def __iter__(self):
+        for index, record in enumerate(self.__rows):
+            yield record
```
Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: Why the numerate? And at this point, it's just a glorified
```python
return iter(self.__rows)
```
so maybe just use that?

File: addons/base_import/odsreader.py [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# 
+import xml.parsers.expat
+
+
+class odsreader:
+    def __init__(self, file):
+        xmldata = file
+
+        self.__iter_cnt = 0
```
Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: remove that

File: addons/base_import/odsreader.py [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+        self.__rows = []
+        self.__working_row = []
+
+        p = xml.parsers.expat.ParserCreate()
+
+        p.StartElementHandler = self.__start_element__
+        p.EndElementHandler = self.__end_element__
+        p.CharacterDataHandler = self.__char_data__
+        # parse the data
+        p.Parse(xmldata, True)
```
Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: Is there a point to aliasing `file` to `xmldata`? Why not use `file` directly? Or call the parameter `xmldata`?

File: addons/base_import/odsreader.py [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+
+    def __iter__(self):
+        for index, record in enumerate(self.__rows):
+            yield record
+
+    def next(self):
+        return
```
Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: I don't think that method is used anymore, remove it.

Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: I don't think that method is used anymore, remove it.

File: addons/base_import/tests/test_cases.py [hunk 33]
```diff
@@ -23,6 +26,12 @@ def sorted_fields(fields):
     recursed = [dict(field, fields=sorted_fields(field['fields'])) for field in fields]
     return sorted(recursed, key=lambda field: field['id'])
 
+def test_module_installed(module_name):
```
Comment (xmo-odoo) [inline addons/base_import/tests/test_cases.py]: Name doesn't really fit, this tests that the module is *not* installed (it returns `None` if the module is installed).

File: addons/base_import/odsreader.py (hunk 31 above)
Comment (xmo-odoo) [inline addons/base_import/odsreader.py]: Is there a point to aliasing `file` to `xmldata`? Why not use `file` directly? Or call the parameter `xmldata`?

**Comment by xmo-odoo** [comment]
I see no way to override the file's content type, did I miss it?

File: addons/base_import/models.py (hunk 16 above)
Comment (xmo-odoo) [inline addons/base_import/models.py]: OK. Reading threads, I've found out 0.9.3 has a `xldate_as_datetime`. Which version are we supposed to support?

File: addons/base_import/models.py (hunk 26 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: Yes, we need `iter` call here, as we need list iterator instead of class object

**Comment by rba-odoo** [comment]
@xmo-odoo resolved your comments, can you have a look?

**Comment by xmo-odoo** [comment]
@rba-odoo there are still comments you haven't replied to or effected.

File: addons/base_import/odsreader.py (hunk 31 above)
Comment (rba-odoo) [inline addons/base_import/odsreader.py]: okay its resolved, I am assigning xmldata by convertinf file into content xml.  

`xmldata = zipfile.ZipFile(StringIO(file)).read('content.xml')`

File: addons/base_import/controllers.py (hunk 3 above)
Comment (rba-odoo) [inline addons/base_import/controllers.py]: @xmo-odoo Done, not require here now.

File: addons/base_import/models.py (hunk 4 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 5 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 6 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 7 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.! , Created file handler function

File: addons/base_import/models.py (hunk 8 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 9 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 10 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done,!

File: addons/base_import/models.py (hunk 12 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/tests/test_cases.py (hunk 13 above)
Comment (rba-odoo) [inline addons/base_import/tests/test_cases.py]: @xmo-odoo Done.! Used `@unittest.skipif`

File: addons/base_import/static/src/js/import.js (hunk 14 above)
Comment (rba-odoo) [inline addons/base_import/static/src/js/import.js]: @xmo-odoo Done.!

File: addons/base_import/static/src/js/import.js (hunk 15 above)
Comment (rba-odoo) [inline addons/base_import/static/src/js/import.js]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 16 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 17 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 18 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/odsreader.py (hunk 19 above)
Comment (rba-odoo) [inline addons/base_import/odsreader.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 20 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 21 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/static/src/js/import.js (hunk 22 above)
Comment (rba-odoo) [inline addons/base_import/static/src/js/import.js]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 23 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 24 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 25 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 27 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/models.py (hunk 28 above)
Comment (rba-odoo) [inline addons/base_import/models.py]: @xmo-odoo Done.!

File: addons/base_import/odsreader.py (hunk 29 above)
Comment (rba-odoo) [inline addons/base_import/odsreader.py]: @xmo-odoo Done.!

File: addons/base_import/odsreader.py (hunk 30 above)
Comment (rba-odoo) [inline addons/base_import/odsreader.py]: @xmo-odoo Done.!

File: addons/base_import/odsreader.py (hunk 32 above)
Comment (rba-odoo) [inline addons/base_import/odsreader.py]: @xmo-odoo Done.!

File: addons/base_import/tests/test_cases.py (hunk 33 above)
Comment (rba-odoo) [inline addons/base_import/tests/test_cases.py]: @xmo-odoo Done.!

File: addons/base_import/models.py [hunk 34]
```diff
@@ -126,6 +142,46 @@ def get_fields(self, cr, uid, model, context=None,
         # TODO: cache on model?
         return fields
 
+    def _file_type_handler(self, file_type, record, options):
+        file_extension = FILE_TYPE_DICT.get(file_type)
+        if file_extension:
+            return getattr(self, '_read_' + file_extension)(record, options)
+        else:
+            raise UserError(_("Not a compatible file format. It should be XLS, XLSX or ODS file"))
```
Comment (matrixise) [inline addons/base_import/models.py]: can you add CSV in your exception.

**Comment by rba-odoo** [comment]
New PR: https://github.com/odoo/odoo/pull/7285

**Pull request closed without merging**